posts/2020/crud-pattern-in-react-implementation.md

```diff
@@ -0,0 +1,130 @@
+  },
+  ...
+}
+
+// Using 'key' as argument to fieldError() and fieldHelperText()
+// to map error and helperText
+<form id="bookForm">
+    <TextField
+        id={formFieldAttributes.title.key}
+        error={this.fieldError(formFieldAttributes.title.key)}
+        helperText={this.fieldHelperText(formFieldAttributes.title.key)}
+        ...
+    />
+...
+</form>
+
+```
+
+`formValidator()` function uses this object to validate form and generate errors (if any). We will check its functioning in the next section.
+
+## Form Validation (`formValidator()` function)
+
+```javascript
+const formValidator = (formFieldAttributes, userInput, isEditForm) => {
+  let formErrors = {};
+  for (let field in formFieldAttributes) {
+    let fieldObj = formFieldAttributes[field];
+    if (isEditForm && !fieldObj.editable) {
+      continue;
+    }
+    let fieldKey = fieldObj.key;
+
+    let value = userInput[fieldKey];
+
+    let fieldEmpty = !value;
+
+    if (fieldEmpty && fieldObj.required) {
+      formErrors[fieldKey] = {
+        error: true,
+        helperText: "this field should be non-empty"
+      };
+    } else if (fieldObj.customValidator && !fieldEmpty) {
+      let customValidator = fieldObj.customValidator;
+      let formError = customValidator(value);
+      if (formError.error) formErrors[fieldKey] = formError;
+    }
+  }
+  return { formErrors, errorCount: Object.keys(formErrors).length };
+};
+```
+
+Form validation is performed by `formValidator()` function (just before an API call is made). This function checks every field value against the attributes mentioned in `formFieldAttributes` object. For every field error, it adds an error object as `{key(fieldName): { error: true/false, helperText: "text"/"" }}` in `formErrors` object. Error object for every field might also be received from `customValidator()`. Finally, this function returns `formErrors` object. The caller will then map errors to field using this `formErrors` object.
+
+In the next section, we will see how errors returned from API response are similary mapped to UI fields.
+
+## API Errors Reporting (`apiErrorsToFormField()` function)
+
+Collecting field specific errors received from API response is done by `apiErrorsToFormFields()` function in very similar manner to that of `formValidator()`. For every field error, an error object (as mentioned in previous section) is added in `apiErrors` object. An error generating pattern can be clearly observed here.
+
+In the next section, we will see how the errors from `formErrors` and `apiErrors` are mapped to the related field.
+
+## Mapping errors to field (`fieldError()` and `FieldHelperText()`)
+
+The errors from `formErrors` and `apiErrors` are combined and mapped to related field by `fieldError()` and `fieldHelperText()` functions. An `id/key` argument is used to check if the field has an error or helperText. Errors from `apiErrors` are mapped only when there are no errors in `formErrors` object.
```

Comment:
This para to be changed to - 

As we have seen `formErrors` and `apiErrors` are similar objects, these objects are used by these helper functions to actually show the errors to the users in the UI. For identifying the UI field to which the errors are to be mapped, the `key` (from the `formFieldAttributes` object) is used here.